OSDOCS-19969: Gateway API backendTLSPolicy for securing HTTPRoutes (#112677)
openshift/openshift-docs: modules/configuring-backend-reencrypt-tls.adoc

[role="_abstract"]
To meet strict security requirements, use re-encrypt TLS termination to encrypt traffic from the gateway to your backend services. This ensures end-to-end encryption while allowing gateway payload inspection to route traffic efficiently.

.Prerequisites
* You have created an edge-terminated listener on your gateway.
* You have created an `HTTPRoute`.
* You have created a target backend service.

[NOTE]
====
Unlike standard {product-title} routes, re-encrypt for `HTTPRoutes` requires a listener configured with edge termination on the gateway and a `BackendTLSPolicy` that targets the service.
====

.Procedure

* Create a `BackendTLSPolicy` custom resource (CR) file that targets your backend service using one of the following configuration options, and then apply it to your cluster:
+
--
** To use system certificates for re-encrypt termination:
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-dev
spec:
  targetRefs:
    - kind: Service
      name: dev
      group: ""
  validation:
    wellKnownCACertificates: "System"
    hostname: dev.example.com

> ----

Reviewer: the backendTLSPolicy should match the namespace as the gateway and the HTTPRoute, it is kind of intuitive, but still a detail which should be mentioned IMO

the hostname should also match the httproute hostname. 

Maybe put these 2 as a note?

Author: Great catch. I've added a note after the the introductory step

** To use explicit CA certificates for re-encrypt termination:
+
Configure the resource to use certificates defined in a `configMap` object, such as `auth-cert` in the following example:
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-auth
spec:
  targetRefs:
    - kind: Service
      name: auth
      group: ""
  validation:
    caCertificateRefs:
      - kind: ConfigMap
        name: auth-cert
        group: ""
    hostname: "authentication.example.com"
    subjectAltNames:
    - type: "URI"
      uri: "spiffe://authorization.example.com/identity"

> ----

Reviewer: The configmap should already be created containing the public (CA) certificate that signed your backend (auth) service's certificate.

Author: Got it, I added this a prerequisite.

+
You must explicitly specify the `sectionName` (the port name) in your `targetRefs` if your target service exposes multiple ports and you want to apply different policies to them.
+
[IMPORTANT]
====
If you omit the `sectionName` when multiple references target the same service, the policy will fail validation or result in a `Conflicted` status.
====
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-multi-port
spec:
  targetRefs:
    - kind: Service
      name: my-service
      group: ""
      sectionName: https
  validation:
    wellKnownCACertificates: "System"
    hostname: my-service.example.com
----
--

.Verification

* Verify that the `BackendTLSPolicy` custom resource was accepted by the implementation controller by checking its status. The following example checks the `tls-upstream-dev` sample policy:
+
[source,terminal]
----
$ oc get backendtlspolicy tls-upstream-dev -o yaml
----
+
.Example output
[source,yaml]
----
status:
  conditions:
  - lastTransitionTime: "2026-06-03T15:16:27Z"
    message: Policy is accepted
    reason: Accepted
    status: "True"
    type: Accepted

> ----

Reviewer: not sure if the message changes, but this is what I get for the message
```
    conditions:
    - lastTransitionTime: "2026-06-05T20:56:41Z"
      message: Configuration is valid
      observedGeneration: 1
      reason: Accepted
      status: "True"
      type: Accepted
```

Author: That's great. Including the exact, real-world phrasing will prevent users from second-guessing whether their configuration actually succeeded. I updated the verification step accordingly

// Module included in the following assemblies:
//
// * networking/ingress_load_balancing/configuring_ingress_cluster_traffic/routing-http-requests-to-services.adoc

:_mod-docs-content-type: PROCEDURE
[id="configuring-backend-reencrypt-tls_{context}"]
= Configuring re-encrypt termination with a BackendTLSPolicy

[role="_abstract"]
To meet strict security requirements, use re-encrypt TLS termination to encrypt traffic from the gateway to your backend services. This ensures end-to-end encryption while allowing gateway payload inspection to route traffic efficiently.

.Prerequisites
* You have created an edge-terminated listener on your gateway.
* You have created an `HTTPRoute`.
* You have created a target backend service.
* If you are using explicit CA certificates, you have created a `ConfigMap` object containing the public CA certificate that signed your backend service's certificate.

[NOTE]
====

> Unlike standard {product-title} routes, re-encrypt for `HTTPRoutes` requires a listener configured with edge termination on the gateway and a `BackendTLSPolicy` that targets the service.

Reviewer 2: ```suggestion
Unlike standard {product-title} routes, re-encrypt for `HTTPRoutes` requires **both** a listener configured with edge termination on the gateway and a `BackendTLSPolicy` that targets the service.
```

Author: Thanks for the suggestion. I've incorporated "both" for clarity. However, I'm not going to bold "both" here because our style guidelines reserve bold text for UI elements and literal values. The admonition provides the necessary visual weight.

====

.Procedure

. Create a `BackendTLSPolicy` custom resource (CR) file that targets your backend service using one of the following configuration options:
+
[NOTE]
====

> The `BackendTLSPolicy` object must reside in the same namespace as your gateway and `HTTPRoute`. Additionally, the `hostname` specified in the policy must match the hostname configured in your `HTTPRoute`.

Reviewer 2: Did @rhamini3 find this to be true?  I'm a bit surprised. Also, none of our examples have the namespace set for BackendTLSPolicy so if if this is true, we need to update this.

Also, hostname is optional, so this note should be reworded.

```suggestion
The `BackendTLSPolicy` object must reside in the same namespace as your gateway and `HTTPRoute`. Additionally, if you choose to specify a `hostname`  in the policy, it must match the hostname configured in your `HTTPRoute`.
```

Reviewer: right, I need to clarify here, the backendTLSPolicy needs to be in the same namespace as the httproute, unless a referenceGrant is configured. It does NOT need to be in the same namespace as the gateway as long as the allowedRoutes.Namespaces is set to `all`.

Author: Thanks, y'all! I've rewritten this note for clarification: the policy must share a namespace with the HTTPRoute (unless using a ReferenceGrant), but doesn't need to share one with the Gateway. 

Also, good catch on the hostname being optional. I've updated that sentence using your suggested phrasing. Last thing, to ensure there is no confusion for the user, I went ahead and added explicit namespace declarations to the BackendTLSPolicy YAML examples.

// Module included in the following assemblies:
//
// * networking/ingress_load_balancing/configuring_gateway_api/securing-httproutes.adoc

:_mod-docs-content-type: PROCEDURE
[id="configuring-backend-reencrypt-tls_{context}"]
= Configuring re-encrypt termination with a BackendTLSPolicy

[role="_abstract"]
To meet strict security requirements, use re-encrypt TLS termination to encrypt traffic from the gateway to your backend services. This ensures end-to-end encryption while allowing gateway payload inspection to route traffic efficiently.

.Prerequisites

> * You created an edge-terminated listener on your gateway.

Reviewer 3: ```suggestion

* You created an edge-terminated listener on your gateway.
```
need a line space after the titles; could cause a migration problem

Author: great catch

// Module included in the following assemblies:
//

> // * nnetworking/ingress_load_balancing/configuring_gateway_api/securing-httproutes.adoc

Reviewer 2: typo ?

```suggestion
// * networking/ingress_load_balancing/configuring_gateway_api/securing-httproutes.adoc
```

Author: Thanks! Copy/paste error. That's just commented-out metadata for other writers who might work on this file in the future but it's still a good fix.

:_mod-docs-content-type: PROCEDURE
[id="configuring-backend-reencrypt-tls_{context}"]
= Configuring re-encrypt termination with a BackendTLSPolicy

[role="_abstract"]
To meet strict security requirements, use re-encrypt TLS termination to encrypt traffic from the gateway to your backend services. This ensures end-to-end encryption while allowing gateway payload inspection to route traffic efficiently.

.Prerequisites

> * You have created an edge-terminated listener on your gateway.

Reviewer 3: ```suggestion

* You created an edge-terminated listener on your gateway.
```
past perfect not needed (also add a line space under section title)

> * You have created an `HTTPRoute`.

Reviewer 3: ```suggestion
* You created an `HTTPRoute`.
```

> * You have created a target backend service.

Reviewer 3: ```suggestion
* You created a target backend service.
```

> * If you are using explicit CA certificates, you have created a `ConfigMap` object containing the public CA certificate that signed your backend service's certificate.

Reviewer 3: ```suggestion
* If you are using explicit certificate authority (CA) certificates, you have created a `ConfigMap` object containing the public CA certificate that signed your backend service's certificate.
```

> [NOTE]

Reviewer 3: ```suggestion
[IMPORTANT]
```

====
Unlike standard {product-title} routes, re-encrypt for `HTTPRoutes` requires both a listener configured with edge termination on the gateway and a `BackendTLSPolicy` that targets the service.
====

.Procedure

> . Create a `BackendTLSPolicy` custom resource (CR) file that targets your backend service using one of the following configuration options:

Reviewer 3: ```suggestion
. Create a `BackendTLSPolicy` custom resource (CR) YAML that targets your backend service. Use one of the following configuration options:
```

+

> [NOTE]

Reviewer 3: this note is not appropriate here and likely will not migrate; place this information after the step or before the procedure starts

Author: thanks, moved it up

====

> The `BackendTLSPolicy` object must reside in the same namespace as your `HTTPRoute` unless you configure a `ReferenceGrant`. It does not need to reside in the same namespace as your gateway, provided the gateway is configured to allow routes from other namespaces. Additionally, if you choose to specify a `hostname` in the policy, it must match the hostname configured in your `HTTPRoute`.

Reviewer 2: I think the "It" in the second sentence needs to be spelled out as an HTTPRoute, not BackendTLSPolicy.

Author: Thanks! To clarify the "it", I've edited the note like this:

"The `BackendTLSPolicy` object must reside in the same namespace as your `HTTPRoute` unless you configure a `ReferenceGrant`. The `HTTPRoute` does not need to reside in the same namespace as your gateway, provided the gateway is configured to allow routes from other namespaces. Additionally, if you choose to specify a `hostname` in the policy, it must match the hostname configured in your `HTTPRoute`."

====
+
--
** To use system certificates for re-encrypt termination:
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-dev
  namespace: my-application
spec:
  targetRefs:
    - kind: Service
      name: dev
      group: ""
  validation:
    wellKnownCACertificates: "System"
    hostname: dev.example.com
----

** To use explicit CA certificates for re-encrypt termination:
+
Configure the resource to use your pre-existing `ConfigMap` object, such as `auth-cert` in the following example:
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-auth
  namespace: my-application
spec:
  targetRefs:
    - kind: Service
      name: auth
      group: ""
  validation:
    caCertificateRefs:
      - kind: ConfigMap
        name: auth-cert
        group: ""
    hostname: "authentication.example.com"
    subjectAltNames:
    - type: "URI"
      uri: "spiffe://authorization.example.com/identity"
----

> 

Reviewer 3: ```suggestion
+
```
might also need one here; check the preview

Author: thanks but this is rendering properly in the preview

** To configure re-encrypt termination for a service with multiple ports:
+

> You must explicitly specify the `sectionName` (the port name) in your `targetRefs` if your target service exposes multiple ports and you want to apply different policies to them.

Reviewer 3: ```suggestion
You must explicitly specify the the port name with the `sectionName` parameter in your `targetRefs` if your target service exposes multiple ports and you want to apply different policies to them.
```
if you want to use both the parameter name and what goes in it, this way is the most common

Author: thanks

+
[IMPORTANT]
====

> If you omit the `sectionName` when multiple references target the same service, the policy will fail validation or result in a `Conflicted` status.

Reviewer 3: ```suggestion
If you omit the `sectionName` when multiple references target the same service, the policy fails validation or result in a `Conflicted` status.
```
no future tense (contrib guide, i believe)

Author: thanks

====
+
[source,yaml]
----
apiVersion: gateway.networking.k8s.io/v1
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-multi-port
  namespace: my-application
spec:
  targetRefs:
    - kind: Service
      name: my-service
      group: ""
      sectionName: https
  validation:
    wellKnownCACertificates: "System"
    hostname: my-service.example.com
----
--

. Apply the policy to your cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f <filename>.yaml
----

.Verification

* Verify that the `BackendTLSPolicy` custom resource was accepted by the implementation controller by checking its status. The following example checks the `tls-upstream-dev` sample policy:
+
[source,terminal]
----
$ oc get backendtlspolicy tls-upstream-dev -n my-application -o yaml
----
+
.Example output
[source,yaml]
----
status:
  conditions:
  - lastTransitionTime: "2026-06-05T20:56:41Z"
    message: Configuration is valid
    observedGeneration: 1
    reason: Accepted
    status: "True"
    type: Accepted
----
+

> If the `Accepted` status is `False`, check the `reason` field. Implementation-specific reasons, such as `InvalidCACertificateRef` or `Pending` (waiting for the implementation controller), indicate why the re-encryption failed.

Reviewer 3: ```suggestion
.Troubleshooting

* If the `Accepted` status is `False`, check the `reason` field. Implementation-specific reasons, such as `InvalidCACertificateRef` or `Pending` show why the re-encryption failed.
```
this is an additional troubleshooting step, not part of the verification IMO (see TaskTitle on the Confluence page for valid Procedure module titles, https://redhat.atlassian.net/wiki/spaces/OSDOCS/pages/265912624/Content+updates+in+preparation+for+migrating+to+DITA)

Author: got it, thanks
